### ISSUE TITLE
something wrong with confidence_loss？

### ISSUE THREAD

**Issue by stezio** [qqwweee/keras-yolo3#360]
@qqwweee 
![image](https://user-images.githubusercontent.com/27137679/53929836-0b612c80-40ca-11e9-8289-38807de71d55.png)
the object confidence should be muliplied by IOU(pred,ground truth) when training as the formulation above from the yolo paper.
It means there should be something done with the object confidence(raw_pred[...,4:5]).
But there is no such calculation found in the code.
Do I miss something?Looking forward to your reply!

**Comment by OtiTantaoui** [comment]
take a look to the function "yolo_boxes_and_scores" in model.py. 
This step was implemented I think in that function (line 202: box_scores = box_confidence * box_class_probs). 
I m not sure as I am just a nob but this is what I understood

**Comment by stezio** [comment]
@OtiTantaoui 
Thanks for your reply!
But 

> function (line 202: box_scores = box_confidence * box_class_probs)

means class_confidence should be multiplied by object_confidence during testing.
It is not the formulation cannot found in the code and is not in the training stage.

**Comment by bluesy7585** [comment]
Does this impacts performance? any one compare trained model performance with darknet?

**Comment by stezio** [comment]
@bluesy7585 
In fact,I found something else different from darknet except this.
And for some reason,this implemention has 1 point lower in mAP compared to the original darknet in my own dataset.

**Comment by HulkMaker** [comment]
> @bluesy7585
> In fact,I found something else different from darknet except this.
> And for some reason,this implemention has 1 point lower in mAP compared to the original darknet in my own dataset.

hi do you use 416x416 image size? And what mAP you got, is 54%?

**Comment by HulkMaker** [comment]
mAP=55.2 (IoU=0.50)  416x416
![image](https://user-images.githubusercontent.com/5285649/56129156-2457da00-5fb4-11e9-9d45-a643129055c1.png)
Anyone got higher than this?

**Comment by githubuse** [comment]
> mAP=55.2 (IoU=0.50) 416x416
> ![image](https://user-images.githubusercontent.com/5285649/56129156-2457da00-5fb4-11e9-9d45-a643129055c1.png)
> Anyone got higher than this?

could you publish you train log,  I found my loss is like this: 
1/1[==========================] -13s 13s/step - loss:1494.9449 - val_loss: 6177.47.17
Epoch 29/50
....
My problem is why my loss is so large, what's the range of loss, i thought it's about in range of 0-1.0.
Thanks

**Comment by AlbertoSabater** [comment]
Can you post the code for the corrected loss function?

**Comment by HulkMaker** [comment]
> > mAP=55.2 (IoU=0.50) 416x416
> > ![image](https://user-images.githubusercontent.com/5285649/56129156-2457da00-5fb4-11e9-9d45-a643129055c1.png)
> > Anyone got higher than this?
> 
> could you publish you train log, I found my loss is like this:
> 1/1[==========================] -13s 13s/step - loss:1494.9449 - val_loss: 6177.47.17
> Epoch 29/50
> ....
> My problem is why my loss is so large, what's the range of loss, i thought it's about in range of 0-1.0.
> Thanks

I just use official weights, not train it. Sorry be not helpful.

**Comment by HulkMaker** [comment]
> Can you post the code for the corrected loss function?

I just use official weights, not train it. Sorry not be helpful.

**Comment by lcltopismine** [comment]
> @OtiTantaoui
> Thanks for your reply!
> But
> 
> > function (line 202: box_scores = box_confidence * box_class_probs)
> 
> means class_confidence should be multiplied by object_confidence during testing.
> It is not the formulation cannot found in the code and is not in the training stage.

Please read yolo v1 carefully. What you refers to defines a confidence score based on real time performance. This is required in test stage for non-max sup. 
This score is not required in the training stage but works to help remove non-necessary box. So the confidential loss is correct.
